### PR TITLE
[confidential-transfer] Add transfer with fee proof generation and extraction

### PR DESCRIPTION
**Pull request by samkim-crypto**
#### Problem
The logic to generate proof data for transfer with fee have not been added to the repo yet.

#### Summary
Add the logic to generate the proof data. This proof is quite technical, but it essentially follows the existing logic in [token client](https://github.com/solana-labs/solana-program-library/blob/master/token/client/src/proof_generation.rs).

### CODE REVIEW

File: token/confidential-transfer/proof-generation/src/transfer_with_fee.rs [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        range_proof_data,
+    ))
+}
+
+fn calculate_fee(transfer_amount: u64, fee_rate_basis_points: u16) -> Option<(u64, u64)> {
+    let numerator = (transfer_amount as u128).checked_mul(fee_rate_basis_points as u128)?;
+
+    // Warning: Division may involve CPU opcodes that have variable execution times.
+    // This non-constant-time execution of the fee calculation can theoretically
+    // reveal information about the transfer amount. For transfers that invole
```
Comment (joncinque) [inline token/confidential-transfer/proof-generation/src/transfer_with_fee.rs:312]: micro-nit
```suggestion
    // reveal information about the transfer amount. For transfers that involve
```

File: token/confidential-transfer/proof-extraction/src/transfer_with_fee.rs [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        &commitment_to_ristretto(transfer_amount_commitment_lo),
+        &commitment_to_ristretto(transfer_amount_commitment_hi),
+    )
+    .ok_or(TokenProofExtractionError::CurveArithmetic)?;
+    let transfer_fee_basis_points_scalar = u16_to_scalar(transfer_fee_basis_points);
+    let scaled_transfer_amount_point =
+        ristretto::multiply_ristretto(&transfer_fee_basis_points_scalar, &transfer_amount_point)
+            .ok_or(TokenProofExtractionError::CurveArithmetic)?;
+
+    const MAX_FEE_BASIS_POINTS: u64 = 10_000;
```
Comment (joncinque) [inline token/confidential-transfer/proof-extraction/src/transfer_with_fee.rs:309]: nit: You can remove this and use the one defined at the top of the file instead

**Review by joncinque — APPROVED** [review]
Looks good! Just tiny things

File: token/confidential-transfer/proof-tests/tests/proof_test.rs [hunk 3]
```diff
@@ -53,3 +57,77 @@ fn test_proof_validity(spendable_balance: u64, transfer_amount: u64) {
...
+fn test_transfer_with_fee_correctness() {
+    test_transfer_with_fee_proof_validity(0, 0, 0, 0);
+    test_transfer_with_fee_proof_validity(0, 0, 0, 1);
+    test_transfer_with_fee_proof_validity(0, 0, 1, 0);
+    test_transfer_with_fee_proof_validity(0, 0, 1, 1);
+    test_transfer_with_fee_proof_validity(1, 0, 0, 0);
+    test_transfer_with_fee_proof_validity(1, 1, 0, 0);
+
+    test_transfer_with_fee_proof_validity(100, 100, 5, 10);
+    test_transfer_with_fee_proof_validity(100, 100, 5, 1);
```
[the remark this answers is not in the document]
Comment (samkim-crypto) [reply token/confidential-transfer/proof-tests/tests/proof_test.rs:71]: No, there is no specific reason. I had a bunch of tests and it seemed like the test was becoming quite slow and the transfer amount was already somewhat tested for transfer without fee tests, so I thought I would just include the smaller ones. Now, I did add the tests for big numbers and used `(5, 10)` and `(5, 1)` for the `(basis point, max fee)` parameters. I can also add more edge case fee parameters too if you see fit.

File: token/confidential-transfer/proof-generation/src/transfer_with_fee.rs [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+    fee_rate_basis_points: u16,
+    maximum_fee: u64,
+) -> Result<
+    (
+        CiphertextCommitmentEqualityProofData,
+        BatchedGroupedCiphertext3HandlesValidityProofData,
+        PercentageWithCapProofData,
+        BatchedGroupedCiphertext2HandlesValidityProofData,
+        BatchedRangeProofU256Data,
+    ),
```
[the remark this answers is not in the document]
Comment (samkim-crypto) [reply token/confidential-transfer/proof-generation/src/transfer_with_fee.rs:52]: Yeah, I think this is a good point and I think we can certainly organize it using a structs. Since this issue is shared with transfer without fee, I'll address it in a separate PR.

File: token/confidential-transfer/proof-extraction/src/transfer_with_fee.rs (hunk 2 above)
Comment (samkim-crypto) [reply token/confidential-transfer/proof-extraction/src/transfer_with_fee.rs]: Yep, thanks for the catch. Fixed!

**Review by joncinque — APPROVED** [review]
